.github/workflows/build.yml

```diff
@@ -0,0 +1,266 @@
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: macos
+          path: build/lmms-*.dmg
+      - name: Print ccache statistics
+        run: |
+          echo "[ccache config]"
+          ccache -p
+          echo "[ccache stats]"
+          ccache -s
+  mingw:
+    strategy:
+      fail-fast: false
+      matrix:
+        arch: ['32', '64']
+    name: mingw${{ matrix.arch }}
+    runs-on: ubuntu-latest
+    container: lmmsci/linux.mingw${{ matrix.arch }}:18.04
+    env:
+      CMAKE_OPTS: >-
+        -DUSE_WERROR=ON
+        -DCMAKE_BUILD_TYPE=RelWithDebInfo
+        -DUSE_COMPILE_CACHE=ON
+      CCACHE_MAXSIZE: 500M
+      MAKEFLAGS: -j2
+    steps:
+      - name: Update and configure Git
+        run: |
+          add-apt-repository ppa:git-core/ppa
+          apt-get update
+          apt-get --yes install git
+          git config --global --add safe.directory /__w/lmms/lmms
+      - name: Check out
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          submodules: recursive
+      - name: Cache ccache data
+        uses: actions/cache@v3
+        with:
+          key: "ccache-${{ github.job }}-${{ matrix.arch }}-${{ github.ref }}\
+            -${{ github.run_id }}"
+          restore-keys: |
+            ccache-${{ github.job }}-${{ matrix.arch }}-${{ github.ref }}-
+            ccache-${{ github.job }}-${{ matrix.arch }}-
+          path: ~/.ccache
+      - name: Configure
+        run: |
+          mkdir build && cd build
+          ../cmake/build_win${{ matrix.arch }}.sh
+      - name: Build
+        run: cmake --build build
+      - name: Build tests
+        run: cmake --build build --target tests
+      - name: Package
+        run: cmake --build build --target package
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: mingw${{ matrix.arch }}
+          path: build/lmms-*.exe
+      - name: Print ccache statistics
+        run: |
+          echo "[ccache config]"
+          ccache -p
+          echo "[ccache stats]"
+          ccache -s
+  msvc:
+    strategy:
+      fail-fast: false
+      matrix:
+        arch: ['x86', 'x64']
+    name: msvc-${{ matrix.arch }}
+    runs-on: windows-2019
+    env:
+      qt-version: '5.15.2'
+    steps:
+      - name: Check out
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          submodules: recursive
+      - name: Cache Qt
+        id: cache-qt
+        uses: actions/cache@v3
+        with:
+          key: qt-${{ env.qt-version }}-msvc-${{ matrix.arch }}-0
+          path: Qt
+      - name: Cache vcpkg dependencies
+        uses: actions/cache@v3
+        with:
+          key: vcpkg-${{ matrix.arch }}-${{ github.ref }}-${{ github.run_id }}
+          restore-keys: |
+            vcpkg-${{ matrix.arch }}-${{ github.ref }}-
+            vcpkg-${{ matrix.arch }}-
+          path: C:\vcpkg\installed
+      - name: Install Qt
```

Comment:
I use [`jurplel/install-qt-action@v3`](https://github.com/jurplel/install-qt-action/tree/v3) in a private project, so this might help.

Reply:
I used that initially, but had some trouble with it, so wrote that small bit of PowerShell instead.

I think the problem was that if you installed both 32- and 64-bit Qt, which we need for 32-bit VST plugins on 64-bit Windows, then it sets `Qt5_DIR` incorrectly (it always gets set to the 32-bit Qt directory, no matter which order you install them), which breaks the build. It should be easy enough to work around, but that was after some earlier frustration where I tried to use a newer version which handles caching itself, but it didn't work because that version hadn't been compiled yet. After multiple issues I just gave up.

I'm hoping to remove the Qt requirement for remote plugins on Windows soon, at which point this should all be much easier and I'll probably switch back to it.

Reply:
> if you installed both 32- and 64-bit Qt, which we need for 32-bit VST plugins on 64-bit Windows, then it sets `Qt5_DIR` incorrectly

https://github.com/jurplel/install-qt-action/blob/d9c92014ea93bab0cd4502c676d5f210d5518eaf/action/src/main.ts#L325 is causing the problem. I think you may work around by installing 64bit Qt first and then installing 32bit Qt with `set-env` set to `false`.

Reply:
Note, you should drop manual Qt caching and rely on the mechanism provided by `jurplel/install-qt-action` if you want to try my suggestion.

Reply:
Ah, I see the newer version of that action with built-in caching has been built since I last tried it. Looking forward to being able to use it.

Reply:
I've tried using that action again but it still seems to be broken. Its `action.yml` tries to run `./action` as a step, which is intended to run the action in that directory within install-qt-action, but instead tries to run the action in that directory within LMMS. See jurplel/install-qt-action#153.

Reply:
I found that the action is broken in several different ways. :sweat_smile: I'm going to report the underlying issues.
In the meantime, you can try using `PhysSong/install-qt-action@v3`

Reply:
Works well now, thanks!